File: nodejs/autonomous/github-trending/src/token-cache.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Simple in-memory token cache for observability tokens.
 * In production, use a more robust caching solution like Redis.
 */

interface CacheEntry {
  token: string;
  expiresAt: number; // Unix ms
}

const EXPIRY_BUFFER_MS = 5 * 60 * 1000; // 5 minutes

const cache = new Map<string, CacheEntry>();

export function cacheToken(agentId: string, tenantId: string, token: string, expiresInMs: number = 60 * 60 * 1000): void {
  const key = `${agentId}:${tenantId}`;
  cache.set(key, {
    token,
    expiresAt: Date.now() + expiresInMs,
  });
  // Avoid logging agentId/tenantId on every cache write to reduce log noise
}

export function getCachedToken(agentId: string, tenantId: string): string | null {
  const key = `${agentId}:${tenantId}`;
  const entry = cache.get(key);

  if (!entry) {
    return null;
  }

  if (Date.now() + EXPIRY_BUFFER_MS >= entry.expiresAt) {
    cache.delete(key);
    return null;
  }

  return entry.token;
}

/**
 * Token resolver called by the A365 Observability exporter when exporting telemetry.
 */
export const tokenResolver = (agentId: string, tenantId: string): string | null => {
  return getCachedToken(agentId, tenantId);
};

File: python/autonomous/github-trending/token_cache.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""Simple in-memory token cache for observability tokens."""

import threading
from datetime import datetime, timedelta, timezone

_lock = threading.Lock()
_cache: dict[str, tuple[str, datetime]] = {}

# Tokens are considered valid if they expire more than 5 minutes from now.
_EXPIRY_BUFFER = timedelta(minutes=5)


def cache_token(agent_id: str, tenant_id: str, token: str, expires_in: timedelta = timedelta(hours=1)) -> None:
    """Cache an observability token for a specific agent/tenant pair."""
    key = f"{agent_id}:{tenant_id}"
    expires_at = datetime.now(timezone.utc) + expires_in
    with _lock:
        _cache[key] = (token, expires_at)


def get_cached_token(agent_id: str, tenant_id: str) -> str | None:
    """Retrieve a cached token if it exists and hasn't expired."""
    key = f"{agent_id}:{tenant_id}"
    with _lock:
        entry = _cache.get(key)
        if entry is None:
            return None
        token, expires_at = entry
        if datetime.now(timezone.utc) + _EXPIRY_BUFFER >= expires_at:
            del _cache[key]
            return None
        return token
